mathics/doc/common_doc.py

```diff
@@ -596,6 +731,220 @@ def __str__(self) -> str:
         return f"== {self.title} ==\n{self.doc}"
 
 
+class DocPart:
+    """
+    Class to gather Part data and its sub-parts (chapters) and format that.
+    """
+
+    def __init__(self, doc, title, is_reference=False):
+        self.doc = doc
+        self.title = title
+        self.chapters = []
+        self.chapters_by_slug = {}
+        self.chapter_class = DocChapter
```

Comment:
Shouldn't be better to make this a class attribute?

Reply:
Yes, there are a lot of opportunities here and other places. 

However this is going to cause a lot more reorganization of the code.  When made a class attribute, all of the corresponding classes will now have to appear  before classes that use them. (Performance gains are marginal)

If you want to do that feel free, but again let's do that after we get things working for Django and LaTeX.

Reply:
@rocky, it is not a matter of performance, but of consistency and clarity: it is weird the need of passing the class of the sub-elements which are known before hand. As I see this, the design pattern that we use in mathics3.format would be clearer, simpler, and less susceptible to bugs. But if the aim is to follow the present design, and if the idea is to improve the design and behavior of the documentation system before the release, I think we need to consider these alternatives.

Reply:
You may misunderstand the code. As currently written, in building the main documentation object we need class constructors for the chapter, and section. These constructors change whether the context is  testing, LaTeX or HTML.  The static assignments in `common_doc.py` are just _default_ values. 

There is probably a way to reorganize everything to reduce this more. But IMHO it would be nice to make another stab at this after release.